src/Microsoft.Health.Fhir.Api/Features/Audit/AuditMiddleware.cs

```diff
@@ -0,0 +1,70 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Net;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Health.Fhir.Api.Features.Routing;
+using Microsoft.Health.Fhir.Core.Features.Context;
+
+namespace Microsoft.Health.Fhir.Api.Features.Audit
+{
+    /// <summary>
+    /// A middleware that logs audit entry that cannot be logged by <see cref="AuditLoggingFilterAttribute"/>.
+    /// </summary>
+    public class AuditMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IFhirRequestContextAccessor _fhirRequestContextAccessor;
+        private readonly IAuditHelper _auditHelper;
+
+        public AuditMiddleware(
+            RequestDelegate next,
+            IFhirRequestContextAccessor fhirRequestContextAccessor,
+            IAuditHelper auditHelper)
+        {
+            EnsureArg.IsNotNull(next, nameof(next));
+            EnsureArg.IsNotNull(fhirRequestContextAccessor, nameof(fhirRequestContextAccessor));
+            EnsureArg.IsNotNull(auditHelper, nameof(auditHelper));
+
+            _next = next;
+            _fhirRequestContextAccessor = fhirRequestContextAccessor;
+            _auditHelper = auditHelper;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            try
+            {
+                await _next(context);
+            }
+            finally
+            {
+                var statusCode = (HttpStatusCode)context.Response.StatusCode;
+
+                // Since authorization filters runs first before any other filters, if the authorization fails,
```

Comment:
Another option would be to remove the OnResultExecuted override from AuditLoggingFilterAttribute and log all of the "executed" audit log here. However, there is a slight problem with that approach. This code relies on MVC framework to have parsed the RouteData already. Because we use the IdentityServer in our testing, when calls are made to the IdentityServer, this middleware is executed but the RouteData is not parsed since the MVC framework is not triggered. We could skip the auditing if "controller" and "action" are not present but we wouldn't know if the middleware is misbehaving. Currently, if the code reaches here and the "controller" or "action" is missing, it will throw AuditException so that we can monitor to make sure the code is behaving the way it should be.

Any thoughts?